swing/src/main/java/wannabe/swing/RenderType.java

```diff
@@ -109,87 +110,390 @@
-        int outTop = r.top + r.vDepth;
-        polygon.addPoint(r.left + r.hDepth, outTop);
-        polygon.addPoint(outRight, outTop);
-        polygon.addPoint(outRight, bottom + r.vDepth);
+        populateAboveRight(r);
       } else {
         // Height indicated by drawing below-right the face.
+        populateBelowRight(r);
+      }
+    }
+  }
+
+  /**
+   * The presence or absence of three neighbors dictate which of eight cases are used to draw:
+   * <ol>
+   * <li>No neighbors
+   * <li>Neighbor to the left
+   * <li>Neighbor above
+   * <li>Neighbor left and above
+   * <li>Neighbor left, above, and above-left
+   * <li>Neighbor left and above-left
+   * <li>Neighbor above and above-left
+   * <li>Neighbor above-left
+   * </ol>
+   * <pre>
+   *  1 22 3  4 55 6  77 8
+   *       3 44 55 66  7  8
+   * </pre>
+   *
+   * TODO might be nice to enumerate all possible points too, and use the same point #s in
+   * each case.
+   * <p>
+   *
+   * Case 1. No blocking neighbor case; 1, 2, and 3 all share points with the face:
+   * <pre>
+   * (5) *--------* (6)
+   *     |         \
+   *     |  *-------* (1)
+   * (4) *  | (2)
+   *      \ |
+   *       \|
+   *        * (3)
+   * </pre>
+   *
+   * Cases 2 and 3. With a single neighbor blocking, the polygon would instead share two
+   * points (1, 2) with the face. 3, and 4 are determined by hdepth and vdepth.
+   * Note that the areas marked with x are omitted because the blocking neighbor will
+   * draw there.
+   * <pre>
+   * (3) *--------* (4)    (4) *\ x
+   *     x\        \           | \
+   *   (2) *--------* (1)      |  * (1)
+   *                       (3) *  |
+   *                            \ |
+   *                             \|
+   *                              * (2)
+   * </pre>
+   *
+   * Cases 4 and 5. With two adjacent blocking neighbors, each neighbor will draw its sides, leaving
+   * no space for this one. If there's a diagonal here, its face will cover the area anyway.
+   *
+   * <p>
+   * Cases 6 and 7. With an adjacent and a diagonal neighbor blocking, the polygon would share
+   * 1 and 2 with the face. 3 and 4 are determined by hdepth and vdepth.  The small rectangle
+   * shown with xx is elided.
+   * <pre>
+   *    (3)
+   *  *--*-----* (4)        *--*
+   *  |xx|      \           |xx|
+   *  *--*-------* (1)  (4) *--* (1)
+   *    (2)             (3) *  |
+   *                         \ |
+   *                          \|
+   *                           * (2)
+   * </pre>
+   *
+   * Case 8. Only diagonal blocking neighbor case; 1, 2, 3, and 6 (same as 2) all share points
+   * with the face:
+   * TODO can this simplify to a 6-point polygon (1, 5, 4, 3, 7, 8, 1)?
+   * <pre>
+   *    (7) *-----* (8)
+   *        |      \
+   * (5) *--*-------* (1)
+   * (4) *  | (2, 6)
+   *      \ |
+   *       \|
+   *        * (3)
+   * </pre>
+   *
+   * Note that there are at most 8 points in any cases.
+   */
+  public void populateAboveLeft(Rendered r) {
```

Comment:
As you can see from the documentation, this is basically a nightmare of manual coding.